File: src/main/java/am/adrian/dungeonkeeper/game/GameMap.java
package am.adrian.dungeonkeeper.game;

import am.adrian.dungeonkeeper.common.object.Creature;
import am.adrian.dungeonkeeper.common.object.GameObject;
import am.adrian.dungeonkeeper.game.tile.UnclaimedPath;
import lombok.Getter;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static am.adrian.dungeonkeeper.helper.CollisionChecker.objectsCollide;

@Getter
public class GameMap {

    private static final Logger logger = LogManager.getLogger(GameMap.class);

    private final int width;
    private final int height;
    private final GameObject[][] objectMap;
    private final List<Creature> creatures;

    public GameMap(int width, int height) {
        this.width = width;
        this.height = height;
        this.objectMap = new GameObject[height][width];
        this.creatures = new ArrayList<>();
        for (int i = 0; i < height; ++i) {
            for (int j = 0; j < width; ++j) {
                objectMap[i][j] = new UnclaimedPath(j, i);
            }
        }
    }

    public void addObject(GameObject object) {
        if (object.getCoords() == null) {
            logger.debug("Not adding object as its coords are null");
            return;
        }
        logger.debug("Adding object with x: {} and y: {}", object.getCoords().getX(), object.getCoords().getY());
        objectMap[object.getCoords().getY()][object.getCoords().getX()] = object;
    }

    public void addObjects(Collection<? extends GameObject> objectsToAdd) {
        objectsToAdd.forEach(this::addObject);
    }

    public void removeObject(GameObject object) {
        if (object.getCoords() == null) {
            logger.debug("Not removing object as its coords are null");
            return;
        }
        removeObject(object.getCoords().getX(), object.getCoords().getY());
    }

    public void removeObject(int x, int y) {
        objectMap[y][x] = new UnclaimedPath(x, y);
    }

    public void addCreature(Creature creature) {
        if (creature.getCoords() == null) {
            logger.debug("Not adding creature as its coords are null");
            return;
        }

        final var x = creature.getCoords().getX();
        final var y = creature.getCoords().getY();

        final var object = objectMap[y][x];

        if (!(object instanceof UnclaimedPath) && objectsCollide(object, creature)) {
            logger.debug("Not adding creature as its colliding with another object");
            return;
        }

        logger.debug("Adding creature with x: {} and y: {}", x, y);
        creatures.add(creature);
    }

    public void addCreature(Collection<? extends Creature> creaturesToAdd) {
        creaturesToAdd.forEach(this::addCreature);
    }

    public void removeCreature(Creature creature) {
        creatures.remove(creature);
    }
}

File: src/main/java/am/adrian/dungeonkeeper/game/GameMap.kt
package am.adrian.dungeonkeeper.game

import am.adrian.dungeonkeeper.helper.CollisionChecker.objectsCollide
import am.adrian.dungeonkeeper.model.gameobject.GameObject
import am.adrian.dungeonkeeper.model.gameobject.character.Creature
import am.adrian.dungeonkeeper.model.gameobject.tile.path.Path
import am.adrian.dungeonkeeper.model.gameobject.tile.path.UnclaimedPath
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class GameMap(
    @Value("\${gameMap.width}") val width: Int,
    @Value("\${gameMap.height}") val height: Int
) {

    val objectMap: Array<Array<GameObject?>> = Array(height) { arrayOfNulls(width) }
    val creatures: MutableList<Creature> = ArrayList()

    init {
        for (i in 0 until height) {
            for (j in 0 until width) {
                objectMap[i][j] = UnclaimedPath(j, i)
            }
        }
    }

    companion object : Logging

    fun addObject(gameObject: GameObject) {
        logger.debug { "Adding object with x: ${gameObject.getCoords().x()} and y: ${gameObject.getCoords().y()}" }
        objectMap[gameObject.getCoords().y()][gameObject.getCoords().x()] = gameObject
    }

    fun addObjects(objectsToAdd: Collection<GameObject>) {
        objectsToAdd.forEach { gameObject: GameObject -> addObject(gameObject) }
    }

    fun removeObject(gameObject: GameObject) {
        logger.debug { "Removing object with x: ${gameObject.getCoords().x()} and y: ${gameObject.getCoords().y()}" }
        removeObject(gameObject.getCoords().x(), gameObject.getCoords().y())
    }

    fun removeObject(x: Int, y: Int) {
        objectMap[y][x] = UnclaimedPath(x, y)
    }

    fun addCreature(creature: Creature) {
        val x: Int = creature.getCoords().x()
        val y: Int = creature.getCoords().y()
        val gameObject: GameObject = objectMap[y][x] ?: throw RuntimeException("Object was null at x: $x and y: $y")
        if (gameObject !is Path && objectsCollide(gameObject, creature)) {
            logger.debug("Not adding creature as its colliding with another object")
            return
        }

        logger.debug { "Adding creature with x: $x and y: $y" }
        creatures.add(creature)
    }

    fun addCreature(creaturesToAdd: Collection<Creature>) {
        creaturesToAdd.forEach { creature: Creature -> addCreature(creature) }
    }

    fun removeCreature(creature: Creature) {
        creatures.remove(creature)
    }
}
